HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+        mNrmMin = self.mNrmMin
+        mNrmMax = self.mNrmMax
+        mNrmCount = self.mNrmCount
+        exp_nest = self.mNrmNestFac
+        # Create grid
+        mNrmGrid = make_grid_exp_mult(
+            ming=mNrmMin, maxg=mNrmMax, ng=mNrmCount, timestonest=exp_nest
+        )
+        # Assign and set it as time invariant
+        self.mNrmGrid = mNrmGrid
+        self.add_to_time_inv("mNrmGrid")
+
+    def initialize_sim(self):
+        """
+        Initialize the state of simulation attributes.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        """
+        RiskyAssetConsumerType.initialize_sim(self)
+        self.state_now["Share"] = np.zeros(self.AgentCount)
+
+    def sim_birth(self, which_agents):
+        """
+        Create new agents to replace ones who have recently died; takes draws of
+        initial aNrm and pLvl, as in ConsIndShockModel, then sets Share, Adjust
+        and post-rebalancing risky asset nNrmTilde to zero as initial values.
+        Parameters
+        ----------
+        which_agents : np.array
+            Boolean array of size AgentCount indicating which agents should be "born".
+
+        Returns
+        -------
+        None
+        """
+
+        RiskyAssetConsumerType.sim_birth(self, which_agents)
+        self.state_now["Share"][which_agents] = 0.0
+        self.state_now["nNrmTilde"][which_agents] = 0.0
+
+    def sim_one_period(self):
+        """
+        Simulates one period for this type.
+        
+        Has to be re-defined instead of using AgentType.sim_one_period() because
+        of the "stages" structure.
+        
+        Parameters
+        ----------
+        None
+        Returns
+        -------
+        None
+        """
+
+        if not hasattr(self, "solution"):
+            raise Exception(
+                "Model instance does not have a solution stored. To simulate, it is necessary"
+                " to run the `solve()` method of the class first."
+            )
+
+        # Mortality adjusts the agent population
+        self.get_mortality()  # Replace some agents with "newborns"
+
+        # Make state_now into state_prev, clearing state_now
+        for var in self.state_now:
+            self.state_prev[var] = self.state_now[var]
+
+            if isinstance(self.state_now[var], np.ndarray):
+                self.state_now[var] = np.empty(self.AgentCount)
+            else:
+                # Probably an aggregate variable. It may be getting set by the Market.
+                pass
+
+        if self.read_shocks:  # If shock histories have been pre-specified, use those
+            self.read_shocks_from_history()
+        else:  # Otherwise, draw shocks as usual according to subclass-specific method
+            self.get_shocks()
+
+        # Sequentially get states and controls of every stage
+        for s in self.stages:
+            self.set_states[s]()
+            self.set_controls[s]()
```

Comment:
If we were very confident in this stages architecture, what we would do is push these changes "upstream" to the core library, so that the AgentType would do this work naturally, and default back to the current normal to maintain backwards compatibility.

I figure you are on a timeline for this project and don't want to resolve the Stages architecture questions entirely before merging this in?

Reply:
I'm not quite on a timeline in the sense that I can go on using the model without it being merged into `master` HARK.

These would be my reasons for merging this in before settling the big-picture Stages question:
- This model has been in the works for about a year and you've mentioned in the past that others would benefit from it being finally made available (the financial-markets folks).
- I don't know when the Stages debate will be settled and, since the code is already public, I'd like to put out [the REMARK](https://github.com/Mv77/RiskyContrib) soon so that people can cite it. It would be nice if the REMARK could use official HARK and not my own branch where this model lives.
- If anything, this could be used as a test case for whatever the final implementation of Stages is introduced. The exercise would be to switch this working-but-unofficial Stage implementation for the official one and see if tests still pass.